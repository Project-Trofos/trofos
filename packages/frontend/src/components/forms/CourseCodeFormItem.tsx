import React from 'react';
import { InfoCircleOutlined } from '@ant-design/icons';
import { Form, Input } from 'antd';

// Renders form elements to input course code

type CourseCodeFormItemProps = {
  disabled?: boolean;
  required?: boolean;
};
export default function CourseCodeFormItem({ disabled, required }: CourseCodeFormItemProps): JSX.Element {
  return (
    <Form.Item
      label="Course Code"
      name="courseCode"
      rules={required ? [{ required: true, message: "Please input your course's code!" }] : []}
      tooltip={{
        title: 'Course code will be used to uniquely identify this course.',
        icon: <InfoCircleOutlined />,
      }}
    >
      <Input disabled={disabled} />
    </Form.Item>
  );
}

CourseCodeFormItem.defaultProps = {
  disabled: false,
  required: false,
};
