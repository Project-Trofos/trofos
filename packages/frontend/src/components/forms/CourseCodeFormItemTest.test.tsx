import { fireEvent, render, screen, waitFor } from '@testing-library/react';
import React from 'react';
import CourseCodeFormItem from './CourseCodeFormItem';
import FormWrapper from './__testutils__/FormRenderHelper.test';

describe('CourseCodeFormItem test', () => {
  function renderForm(required?: boolean) {
    return render(
      <FormWrapper>
        <CourseCodeFormItem required={required} />
      </FormWrapper>,
    );
  }

  describe('when rendering', () => {
    it('should have correct fields', () => {
      const { baseElement } = renderForm();
      expect(baseElement).toMatchSnapshot();

      expect(screen.getByLabelText(/course code/i)).toBeInTheDocument();
    });
  });

  describe('when code is required', () => {
    it('should disallow submission', async () => {
      const { baseElement } = renderForm(true);

      // eslint-disable-next-line testing-library/no-node-access
      expect(baseElement.getElementsByClassName('ant-form-item-required').length).toBe(1);

      fireEvent.click(screen.getByText(/submit/i));

      await screen.findByText('Please input', { exact: false });

      // Required prompt disappears upon input
      fireEvent.change(screen.getByLabelText(/course code/i), { target: { value: 'CS3203' } });
      await waitFor(() => expect(screen.queryByText('Please input', { exact: false })).toBeNull());
    });
  });

  describe('when code is not required', () => {
    it('should allow submission', async () => {
      const { baseElement } = renderForm(false);

      // eslint-disable-next-line testing-library/no-node-access
      expect(baseElement.getElementsByClassName('ant-form-item-required').length).toBe(0);
    });
  });
});
