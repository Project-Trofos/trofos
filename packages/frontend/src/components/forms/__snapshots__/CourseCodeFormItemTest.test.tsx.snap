// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CourseCodeFormItem test when rendering should have correct fields 1`] = `
<body>
  <div>
    <form
      class="ant-form ant-form-horizontal"
    >
      <div
        class="ant-form-item"
      >
        <div
          class="ant-row ant-form-item-row"
        >
          <div
            class="ant-col ant-form-item-label"
          >
            <label
              class=""
              for="courseCode"
              title="Course Code"
            >
              Course Code
              <span
                aria-label="info-circle"
                class="anticon anticon-info-circle ant-form-item-tooltip"
                role="img"
                title=""
              >
                <svg
                  aria-hidden="true"
                  data-icon="info-circle"
                  fill="currentColor"
                  focusable="false"
                  height="1em"
                  viewBox="64 64 896 896"
                  width="1em"
                >
                  <path
                    d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm0 820c-205.4 0-372-166.6-372-372s166.6-372 372-372 372 166.6 372 372-166.6 372-372 372z"
                  />
                  <path
                    d="M464 336a48 48 0 1096 0 48 48 0 10-96 0zm72 112h-48c-4.4 0-8 3.6-8 8v272c0 4.4 3.6 8 8 8h48c4.4 0 8-3.6 8-8V456c0-4.4-3.6-8-8-8z"
                  />
                </svg>
              </span>
            </label>
          </div>
          <div
            class="ant-col ant-form-item-control"
          >
            <div
              class="ant-form-item-control-input"
            >
              <div
                class="ant-form-item-control-input-content"
              >
                <input
                  class="ant-input"
                  id="courseCode"
                  type="text"
                  value=""
                />
              </div>
            </div>
          </div>
        </div>
      </div>
      <button
        class="ant-btn ant-btn-default"
        type="submit"
      >
        <span>
          submit
        </span>
      </button>
    </form>
  </div>
</body>
`;
